feat: use latest Amplitude Browser 2 SDK (2.11.10) for flutter SDK web

chungdaniel opened the pull request:
Primary changes:
- Update web files for using Browser SDK 2 as opposed to maintenance Javascript SDK
  - Update web bridge file
  - Update dart-js interface file
  - Update example app to use new script loader
  - Add web-specific configuration options
  - Session replay, autocapture are not currently supported (autocapture functionality to be investigated later)
- Use `js_interop` and `js_interop_unsafe` rather than `package:js/js.dart` and `package:js/js_util.dart`
  - Added minimum flutter version of 3.3 to dependency matrix to support `js_interop`


Extras: 
- Improved documentation for members of Configuration in configuration.dart
- Fixed lint issues/linter github actions workflow

Note:
- No tests have been added for web since dart tests default to the Dart VM platform which does not have access to js libraries (e.g. `js_interop`, `package:js/js.dart`)


Jira tickets:
- Update web bridge https://amplitude.atlassian.net/browse/AMP-89343
- Update dart-js interface file https://amplitude.atlassian.net/browse/AMP-89344
- Migrate to js_interop https://amplitude.atlassian.net/browse/AMP-104922

chungdaniel committed:
WIP: have updated sdk able to track events, using js_interop_unsafe

chungdaniel committed:
WIP: Add a few methods to handleMethodCall

chungdaniel committed:
WIP: able to initialize amplitude without errors

chungdaniel committed:
WIP: have created classes for js_interop for configuration, event
Library plugin not yet functional, need to instantiate before passing
Autocapture seems to be occuring when using unified script

chungdaniel committed:
WIP: removed js_interop classes as they were redundant to dart classes
- Use js_interop_unsafe to map directly to JSObject for interop
- Use script loader to load Amplitude-Typescript as unified script had bugs
- Add some web specific configurations to configuration.dart

chungdaniel committed:
chore: Update documentation for running example app for web

chungdaniel committed:
chore: remove comments, add helper method

chungdaniel committed:
chore: update min flutter version in README

izaaz commented on example/pubspec.yaml:
```diff
@@ -3,7 +3,7 @@ description: Demonstrates how to use the amplitude_flutter plugin.
 
 environment:
-  sdk: ">=2.12.0 <3.0.0"
+  sdk: ">=3.3.0 < 4.0.0"
```
Why 3.3 and not 3.0?

chungdaniel replied:
Looking at the changelog, I see `js_interop` as early as 3.1, but there some breaking changes between 3.1 and 3.3, the most important one being [extension types ](https://dart.dev/language/extension-types). [This blogpost](https://medium.com/dartlang/history-of-js-interop-in-dart-98b06991158f) from dartlang says extension types and js_interop were officially released with 3.3 and are what enable WASM with flutter (main driving factor for ditching `package:js/js.dart`).

izaaz commented on lib/amplitude_web.dart:
```diff
@@ -191,77 +76,56 @@ class AmplitudeFlutterPlugin {
+    var configuration = mapToJSObj(call.arguments);
+
+    // autocapture is not supported in flutter web
+    configuration.setProperty('autocapture'.toJS, false.toJS);
```
Are we not doing any autocapture including page views?

chungdaniel replied:
Will need to do some more testing on this, but wanted a first pass on the code review. Will need to build out the example app a bit more to have another page to test out page views, file downloads, etc.

chungdaniel replied:
I have [created a ticket](https://amplitude.atlassian.net/browse/AMP-121410) for following up on this, but for first go at web SDK will release without autocapture

izaaz commented on lib/configuration.dart:
```diff
@@ -34,6 +36,24 @@ class Configuration {
   /// Web specific
   String? appVersion;
+  /// Web specific
+  String? deviceId;
```
device Id can be set in iOS and Android too right? Similary app version too. 

Can we group config by platform?

Cross platform first, then specific platforms next?

chungdaniel replied:
Yep, though in iOS and Android it's only set via `setDeviceId()` according to docs. Web can do this as well, though there is also the option to do so through configuration at initialization time. I don't see anything regarding app version in iOS/Android docs though (but do see for userId, which is a similar story to deviceId).

I've grouped the configs!

izaaz replied:
Android config seems to take in device id. I thought iOS did that too. Its quite possible the docs are not up to date.

Mercy811 replied:
@izaaz it's already group by platforms?

izaaz commented on lib/constants.dart:
```diff
@@ -12,6 +12,7 @@ class Constants {
   static const logLevel = 'info';
   static const flushMaxRetries = 5;
   static const minTimeBetweenSessionsMillis = 5 * 60 * 1000; // 5 minutes
+  static const sessionTimeoutMillis = 30 * 60 * 1000; // 30 minutes
```
whats the difference between minTimeBetweenSessionsMillis and sessionTimeoutMillis?

chungdaniel replied:
They are named differently and have different default values in mobile and web SDKs, but seem to serve the same purpose of session timeout. For the flutter SDK we could simply use the preexisting `minTimeBetweenSessionsMillis` and map it to `sessionTimeout` in the `toMap` function as well to satisfy the browser SDK if desired. What are your thoughts?

izaaz replied:
if these are constants and they serve the same config in two different platforms, then maybe we should call them `minTimeBetweenSessionsMillisForMobile` and `minTimeBetweenSessionMillisForWeb`

When a timeout is not configured in the config, we can use these default values, and when a timeout is configured, we set the same timeout across 3 different platforms. If a customer wants to use two different timeouts for mobile and web, we can implement that when someone asks for it.

chungdaniel committed:
Add new configuration members to .toMap and use []= operator

chungdaniel committed:
Group configuration by platform

izaaz commented on lib/configuration.dart:
```diff
@@ -15,25 +17,33 @@ class Configuration {
+
   /// Web specific
   String? appVersion;
+  String? deviceId;
```
Can we also add a comment for each of these configs so that it shows up in autocomplete when someone tries to use this?

izaaz commented on lib/configuration.dart:
```diff
@@ -15,25 +17,33 @@ class Configuration {
+  int sessionTimeout;
+  String? userId;
+  String? transport;
+  bool offline;
```
I remember iOS and Android too have some offline mode options?

chungdaniel replied:
I've moved this to all platforms and added `offline` to Android/iOS configuration mapping functions.

izaaz commented on lib/amplitude_web.dart:
```diff
@@ -191,77 +76,56 @@ class AmplitudeFlutterPlugin {
-    Identify identify = new Identify();
-    userProperties.forEach((String operation, dynamic properties) {
-      properties.forEach((String key, dynamic value) {
-        switch (operation) {
```
Don't we need all these operations anymore?

chungdaniel replied:
These are to be handled by users in their implementations after #166 and [#168](https://github.com/amplitude/Amplitude-Flutter/pull/168/files#diff-f705947ca59439dd16cdd44ef05768d6c969b469698761059bd514fe4d308203). 

E.g. `identify_form.dart` provides example usage
```
    final Identify identify = Identify()
      ..set('identify_test', 'identify sent at ${DateTime
          .now()
          .millisecondsSinceEpoch}')
    ..add('identify_count', 1);

    if (userPropKey.isNotEmpty && userPropValue.isNotEmpty) {
      identify.set(userPropKey, userPropValue);
    }

    AppState.of(context)
      ..analytics.identify(identify)
      ..setMessage('Identify sent.');
```
Then in `amplitude.dart` we create an identify event that is passed to `amplitude_web.dart` where we simply call `track()` with the event.


See also: https://github.com/amplitude/Amplitude-Flutter/pull/216#discussion_r1901394799

Mercy811 commented on lib/amplitude_web.dart:
```diff
@@ -191,77 +76,56 @@ class AmplitudeFlutterPlugin {
+  /// method.
+  ///
+  /// Returns:
+  /// - `JSObject`: A JavaScript object representing the event properties.
```
Nit: Looks like `getEvent()` returns an event rather than event properties based on how it's used
```
JSObject event = getEvent(call);
amplitude.track(event); 
```

Mercy811 commented on lib/configuration.dart:
```diff
@@ -15,25 +17,33 @@ class Configuration {
-  /// Mobile (iOS and Android) specific
   int identifyBatchIntervalMillis;
-  /// Mobile (iOS and Android) specific
   bool migrateLegacyData;
```
Can we keep the comment for each option? It's written intentionally so that IDE will pop up the comment when DEV use it

Mercy811 commented on lib/configuration.dart:
```diff
@@ -66,8 +76,18 @@ class Configuration {
-  }): trackingOptions = trackingOptions ?? TrackingOptions() {
+    this.deviceId,
+    CookieOptions? cookieOptions,
+    String identityStorage = '',
```
why setting it as an empty string? Browser SDK defaults it to `'cookie'`

chungdaniel replied:
I defined an enum `IdentityStorage` for possible values, and initialized it to `IdentityStorage.cookie.name` in constructor code (since you can't use extension methods in constant expressions here on line 81), though I'm realizing now that this won't show up in IDE intellisense. I'll change this to 'cookie'.

Mercy811 commented on lib/web/amplitude_js.dart:
```diff
@@ -1,63 +1,13 @@
+  external void setUserId(JSString userId);
+  external void setDeviceId(JSString devideId);
+  external void reset();
+  external void flush();
```
Missing identify(), groupIdentify(), revenue()...?

chungdaniel replied:
Are identify/groupIdentify/revenue/setGroup just events with certain properties set? I noticed that for Flutter 4.x, iOS and Android have these properties set in `amplitude.dart`, but when they're received in the respective .kt or .swift bridge files, they all simply call `track` with merged properties on an instance of `Event`. I copied this behaviour for web implementation.

Could you verify if these raw event JSONs look as expected for calling those methods?

identify():
```
{
 "time": 1735862153646,
 "device_id": "ce9ecb87-1dfb-45cc-83ec-2f8031a00606",
 "session_id": 1735862114304,
 "platform": "Web",
 "language": "en-US",
 "ip": "$remote",
 "insert_id": "3f2a18d6-4d05-4be3-89e5-99da6012b7fc",
 "partner_id": null,
 "event_type": "$identify",
 "user_properties": {
  "$set": {
   "identify_test": "identify sent at 1735862153605",
   "identify": "identify"
  },
  "$add": {
   "identify_count": 1
  }
 },
 "attempts": 0,
 "event_id": 0,
 "library": "amplitude-flutter/4.0.0-beta.5_amplitude-ts/2.11.10",
 "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
 "responseCode": 400
}
```

setGroup():
```
{
 "time": 1735862154194,
 "device_id": "ce9ecb87-1dfb-45cc-83ec-2f8031a00606",
 "session_id": 1735862114304,
 "platform": "Web",
 "language": "en-US",
 "ip": "$remote",
 "insert_id": "8ff03106-044c-4dcd-b2ca-8c30d33d4c66",
 "partner_id": null,
 "event_type": "$identify",
 "user_properties": {
  "$set": {
   "group/account": "group/account"
  }
 },
 "groups": {
  "group/account": "group/account"
 },
 "attempts": 0,
 "event_id": 1,
 "library": "amplitude-flutter/4.0.0-beta.5_amplitude-ts/2.11.10",
 "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
 "responseCode": 400
}
```

groupIdentify():
```
{
 "time": 1735862155295,
 "device_id": "ce9ecb87-1dfb-45cc-83ec-2f8031a00606",
 "session_id": 1735862114304,
 "platform": "Web",
 "language": "en-US",
 "ip": "$remote",
 "insert_id": "c58b2453-7c2f-43f8-a07b-849f638c71f4",
 "partner_id": null,
 "event_type": "$groupidentify",
 "groups": {
  "group/account": "group/account"
 },
 "group_properties": {
  "$set": {
   "groupidentify": "groupidentify"
  }
 },
 "attempts": 0,
 "event_id": 2,
 "library": "amplitude-flutter/4.0.0-beta.5_amplitude-ts/2.11.10",
 "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
 "responseCode": 400
}
```

revenue():
```
{
 "time": 1735862156145,
 "device_id": "ce9ecb87-1dfb-45cc-83ec-2f8031a00606",
 "session_id": 1735862114304,
 "platform": "Web",
 "language": "en-US",
 "ip": "$remote",
 "insert_id": "a094fea9-e6d2-4641-bae5-bd535b572a6f",
 "partner_id": null,
 "event_type": "revenue_amount",
 "event_properties": {
  "$price": 41.23,
  "$quantity": 2,
  "$productId": "specialProduct"
 },
 "attempts": 0,
 "event_id": 3,
 "library": "amplitude-flutter/4.0.0-beta.5_amplitude-ts/2.11.10",
 "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
 "responseCode": 400
}
```

chungdaniel replied:
They seem to match what I'm seeing in from the example app for the browser SDK though I'm uncertain if there's more to it than that.

Mercy811 replied:
Oh I see. You're right. `amplitude_js.dart` declares web sdk interfaces. All identify, groupIdentify, etc can be converted to track() so that they are not declared here. Payload looks good!

Mercy811 reviewed (changes requested):
Thanks @chungdaniel for supporting Browser 2.x in Flutter. Overall looks good to me. See comments. Can we add update the [test](https://github.com/amplitude/Amplitude-Flutter/blob/beta/test/configuration_test.dart) for the new configuration options? Thanks

chungdaniel committed:
feat: add optOut for browser SDK
- also change wording for getEvent for PR comment

chungdaniel committed:
fix: Address PR comments:
- Add device ID doc for Android
- Make documentation more explicit to show in editor on hover
- Use single `minTimeBetweenSessionsMillis` argument to map to both mobile and web
- Add "offline" to configuration for iOS/Android

chungdaniel committed:
chore: reword doc for getEvent

chungdaniel committed:
chore: fix lint issues

chungdaniel committed:
chore: add new configuration options to test
also rename minTimeBetweenSessionsMillisForWeb for consistency

chungdaniel committed:
chore: add new config to test

chungdaniel committed:
Update swift-lint.yml

Lint was failing as maxos-12 has been deprecated https://github.com/amplitude/Amplitude-Flutter/actions/runs/12656972512/job/35270750527

chungdaniel committed:
Use macos-13

swiftlint not found, potentially related to xcode version

chungdaniel committed:
Update swift-lint.yml

chungdaniel committed:
remove test key

Mercy811 reviewed (approved):
Thank you Daniel! LGTM

chungdaniel commented:
I've decided to remove `offline` from the config for all platforms as the docs seem inconsistent. I've made a ticket to revisit this in the future. https://amplitude.atlassian.net/browse/AMP-121543

chungdaniel committed:
remove offline config for all platforms
